stablehlo/conversions/tosa/transforms/StablehloQuantLegalizeToTosaRescale.cpp

```diff
@@ -0,0 +1,257 @@
+
+  return buildRescale(rewriter, loc, outputType, inputVal, multiplier, shift,
+                      /* input_zp = */ 0, outputZp,
+                      /* double_round = */ false,
+                      /* scale32 = */ true,
+                      /* per_channel = */ false);
+}
+
+template <typename StablehloOp>
+LogicalResult matchAndRewriteAddSub(StablehloOp op, PatternRewriter& rewriter) {
+  Value lhs = op.getLhs();
+  Value rhs = op.getRhs();
+  Value result = op.getResult();
+
+  auto lhsType = dyn_cast<mlir::ShapedType>(lhs.getType());
+  auto rhsType = dyn_cast<mlir::ShapedType>(rhs.getType());
+  auto resultType = dyn_cast<mlir::ShapedType>(result.getType());
+
+  if (!lhsType || !rhsType || !resultType) {
+    return rewriter.notifyMatchFailure(
+        op, "inputs and result tensors should all have shaped tensor types");
+  }
+
+  auto lhsQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(lhsType.getElementType());
+  auto rhsQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(rhsType.getElementType());
+  auto resultQType =
+      dyn_cast<mlir::quant::UniformQuantizedType>(resultType.getElementType());
+
+  if (!lhsQType || !rhsQType || !resultQType) {
+    if (lhsQType || rhsQType || resultQType) {
+      return rewriter.notifyMatchFailure(op,
+                                         "inputs and result tensors should be "
+                                         "all quantized or all un-quantized");
+    }
+    // not quantized - skip
+    return failure();
+  }
+
+  // lhs/rhs/result are all quantized
+
+  // Following quantization described in tensorflow/lite/kernels/add.cc
+  // In details it does:
+  // 1. Rescale inputs to scale = 2.0 x max(lhs.scale, rhs.scale)
+  // 2. Extra left shift to input to increase precision
+  // Where input_shift = 20 if input is 8-bit
+  // input_shift = 15 if input is 16-bit
+
+  double lhsScale = lhsQType.getScale();
+  double rhsScale = rhsQType.getScale();
+  double resultScale = resultQType.getScale();
+  double maxScale2x = 2.0 * std::max(lhsScale, rhsScale);
+
+  const int32_t SHIFT_8_BIT = 20;
+  const int32_t SHIFT_16_BIT = 15;
+
+  int32_t inputShift = (resultQType.getStorageTypeIntegralWidth() == 16)
+                           ? SHIFT_16_BIT
+                           : SHIFT_8_BIT;
+
+  double lhsRescaleScale = lhsScale / maxScale2x;
+  double rhsRescaleScale = rhsScale / maxScale2x;
+  double resultRescaleScale =
+      maxScale2x / (resultScale * static_cast<double>(1 << inputShift));
+
+  auto loc = op.getLoc();
+
+  // implement single rounding only
+  Value rescaledLhs = buildRescaleToInt32(
+      rewriter, loc, lhs,
+      /* scale = */ lhsRescaleScale * static_cast<double>(1 << inputShift),
+      lhsQType.getZeroPoint());
+  Value rescaledRhs = buildRescaleToInt32(
+      rewriter, loc, rhs,
+      /* scale = */ rhsRescaleScale * static_cast<double>(1 << inputShift),
+      rhsQType.getZeroPoint());
+  if (!rescaledLhs || !rescaledRhs) {
```

Comment:
remove the check if they cannot return nullptr

Reply:
done